File: templates/template.py
import sys
from typing import List


def read_data(file_name: str) -> str:
    data: str
    with open(file_name, "r") as file:
        data = file.read()
    return data


# TODO
def solve1(data):
    pass


# TODO
def solve2(data):
    pass


if __name__ == "__main__":
    data: str
    input_data: str = "data.txt"
    solution_data: str = "solution.txt"
    if len(sys.argv) == 1:
        input_data = "./"+input_data
        solution_data = "./"+solution_data
    elif len(sys.argv) == 2:
        execution_type = sys.argv[1]
        if execution_type.lower().strip() == "test":
            input_data = "./test_" + input_data
            solution_data = "./test_" + solution_data
        elif execution_type.lower().strip() == "full":
            input_data = "./"+input_data
            solution_data = "./"+solution_data
        else:
            print(f"Usage: python3 {sys.argv[0]} [test|full]", file=sys.stderr)
            exit(1)
    elif len(sys.argv) > 2:
        print(f"Usage: python3 {sys.argv[0]} [test|full]", file=sys.stderr)
        exit(1)

    data: str = read_data(input_data)
    solutions: (int, int)
    with open(solution_data) as file:
        solutions = tuple(int(sol) for sol in file.readline().strip().split())

    result1 = solve1(data)
    print(result1)
    assert result1 == solutions[0]
    if len(solutions) > 1:
        result2 = solve2(data)
        print(result2)
        assert result2 == solutions[1]

File: templates/template.c
#include <assert.h>
#include <fcntl.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>

#define BUFFER_SIZE 128

typedef struct {
    int32_t first;
    int32_t second;
} SolutionPair, *PSolutionPair;

SolutionPair get_solution(char *solution_data_file_name);
size_t solve1(const char *const data, const size_t size);
size_t solve2(const char *const data, const size_t size);

SolutionPair get_solution(char *solution_data_file_name) {
    SolutionPair result = {0};

    int solution_data_file = open(solution_data_file_name, O_RDONLY);
    struct stat stat;
    int err = fstat(solution_data_file, &stat);

    char *data = (char *)malloc(sizeof(char) * stat.st_size);
    read(solution_data_file, data, stat.st_size);

    char *endptr;

    result.first = strtoll(data, &endptr, 10);
    result.second = strtoll(endptr, &endptr, 10);

    free(data);
    return result;
}

// TODO
size_t solve1(const char *const data, const size_t size) { return 0; }

// TODO
size_t solve2(const char *const data, const size_t size) { return 0; }

int main(int argc, char **argv) {
    char input_data_file_name[BUFFER_SIZE] = "data.txt";
    char solution_data_file_name[BUFFER_SIZE] = "solution.txt";
    char buf[BUFFER_SIZE] = {0};
    if (argc == 1) {
        snprintf(buf, BUFFER_SIZE - 1, "%s%s", "./", input_data_file_name);
        strncpy(input_data_file_name, buf, BUFFER_SIZE - 1);
        snprintf(buf, BUFFER_SIZE - 1, "%s%s", "./", solution_data_file_name);
        strncpy(solution_data_file_name, buf, BUFFER_SIZE - 1);
    } else if (argc == 2) {
        char *execution_type = argv[1];
        if (!strncasecmp(execution_type, "test", 5)) {
            snprintf(buf, BUFFER_SIZE - 1, "%s%s", "./test_",
                     input_data_file_name);
            strncpy(input_data_file_name, buf, BUFFER_SIZE - 1);
            snprintf(buf, BUFFER_SIZE - 1, "%s%s", "./test_",
                     solution_data_file_name);
            strncpy(solution_data_file_name, buf, BUFFER_SIZE - 1);
        } else if (!strncasecmp(execution_type, "full", 5)) {
            snprintf(buf, BUFFER_SIZE - 1, "%s%s", "./", input_data_file_name);
            strncpy(input_data_file_name, buf, BUFFER_SIZE - 1);
            snprintf(buf, BUFFER_SIZE - 1, "%s%s", "./",
                     solution_data_file_name);
            strncpy(solution_data_file_name, buf, BUFFER_SIZE - 1);
        } else {
            printf("Usage: %s [test|full]\n", argv[0]);
            exit(1);
        }
    } else if (argc > 2) {
        printf("Usage: %s [test|full]\n", argv[0]);
        exit(1);
    }

    SolutionPair solutions = get_solution(solution_data_file_name);

    int input_data_file = open(input_data_file_name, O_RDONLY);
    struct stat stat;
    fstat(input_data_file, &stat);

    char *input_data =
        mmap(NULL, stat.st_size, PROT_READ, MAP_PRIVATE, input_data_file, 0);

    size_t result1 = solve1(input_data, stat.st_size);
    printf("%ld\n", result1);
    assert(result1 == solutions.first);

    size_t result2 = solve2(input_data, stat.st_size);
    printf("%ld\n", result2);
    assert(result2 == solutions.second);

    munmap(input_data, stat.st_size);
    return 0;
}
